ntex-bytes/src/buf/pool.rs

```diff
@@ -0,0 +1,269 @@
+        (**self).take(len)
+    }
+}
+
+/// A sink of buffers. Dropped buffers will automatically return to this sink.
+pub trait BufferSink {
+    /// The type of buffer returned by this buffer source that owns its contents.
+    type Owned: Owned;
+
+    /// The type of backing buffer that is returned to this pool when the last owned or shared buffer
+    /// holding on to said backing buffer is dropped.
+    type Backing;
+
+    /// Returns the given backing buffer to the pool. This is intended to be invoked from
+    /// the `Drop` impls of this pool's owned and shared buffer types.
+    fn put_back(&self, backing: Self::Backing);
+}
+
+/// Convenience trait for a type that is a buffer pool, ie a type that is both a [`BufferSource`]
+/// and a [`BufferSink`].
+pub trait BufferPool: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferPool for T where T: BufferSource + BufferSink<Owned = <Self as BufferSource>::Owned> {}
+
+impl<T> BufferSink for Rc<T>
+where
+    T: ?Sized + BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+impl<T> BufferSink for Arc<T>
+where
+    T: BufferSink,
+{
+    type Backing = T::Backing;
+    type Owned = T::Owned;
+
+    fn put_back(&self, backing: Self::Backing) {
+        (&**self).put_back(backing);
+    }
+}
+
+/// A [`BufferSource`] that always yields new `BytesMut` without blocking.
+#[derive(Clone, Copy, Debug)]
+pub struct BytesPool;
+
+impl BufferSource for BytesPool {
+    type Owned = BytesMut;
+    type Accumulator = BytesAccumulator;
+    type TakeFuture = future::Ready<Self::Owned>;
+
+    fn take(&self, len: usize) -> Self::TakeFuture {
+        future::ready(BytesMut::with_capacity(len))
+    }
+}
+
+impl BufferSink for BytesPool {
+    type Backing = ();
+    type Owned = <Self as BufferSource>::Owned;
+
+    fn put_back(&self, _backing: Self::Backing) {
+    }
+}
+
+/// A buffer that owns a particular range of its backing buffer.
+///
+/// An `Owned` tracks what part of itself has been filled with data.
+/// The filled region can be accessed with [`Owned::filled`].
+/// The unfilled region can be accessed with [`Owned::unfilled_mut`], and bytes can be moved from the start of this region
+/// into the end of the filled region with [`Owned::fill`].
+///
+/// An `Owned` can be subdivided into smaller `Owned`s with `split_to` that each own
+/// smaller splits of the backing buffer.
+///
+/// An `Owned` is not `Clone`. It can be converted to a [`Shared`] which is, via [`Owned::freeze`]
+pub trait Owned {
+    /// The type of shared buffers with the same backing buffer.
+    type Shared: Shared;
+
+    /// Try to append a `u8` to this buffer's filled region.
+    ///
+    /// Returns `None` if there is no space in the filled region to append this `u8`.
+    fn try_put_u8(&mut self, n: u8) -> Option<()> {
```

Comment:
I don't like Option<()> as a result, why can't we just use bool?

Reply:
`Option` is easier to convert into an `Err()` when it's `None`, via `.ok_or{,_else}`. [Example](https://github.com/Azure/iotedge-broker/blob/46147f14c2dcc775bc321ca169ac67b1d6f20d42/data-plane/mqtt-proto/src/v5/connect.rs#L213-L217) (`ByteBuf` there is equivalent to this PR's `BytesAccumulator`).

If it was `bool`, it would need to be clunkier: `.try_put_u8().then(|| ()).ok_or(err)?`.

`-> Option<()>` does have the disadvantage that it's weird for a `try_*` function to return `Option` instead of `Result`. So it could instead be `-> Result<(), ()>`, for which there *is* precedent in a popular crate with [`url::Url::set_port`](https://docs.rs/url/2.2.2/url/struct.Url.html#method.set_port) (and a few other functions). I don't think I like `Result` with `E = ()`, but if you prefer that over `Option<()>` I can switch to it.

Reply:
lets use bool